mysql80: init at 8.0.17

totten opened the pull request:
###### Motivation for this change

MySQL 8.0 is a new shiny. Need mysql80 binaries so that we can start testing our app with it.

Related links:

* [MySQL 8.0 Release Notes](https://dev.mysql.com/doc/relnotes/mysql/8.0/en/)
* [What's New in MySQL 8.0](https://www.mysql.com/why-mysql/white-papers/whats-new-mysql-8-0/)

The PR introduces version ~~8.0.16~~ 8.0.17.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

###### Other comments

This is my first submission to nixpkgs. It's just an adaptation of the `mysql57` scripts, with updates for compiler compatibility.

I've done *very light* testing (manually launching `mysqld`; using `mysql` to create a small database+table).

totten committed:
mysql80: init 8.0.16

MySQL 8.0 is a significant iteration after MySQL 5.7.  This patch adds it as
a parallel build alongside mysql57 (similar to mysql56 and mysql55 before).

aanderse commented:
@totten Thanks for deciding to contribute to nixpkgs! It looks like `8.0.17` has been released so when you get a chance would you mind rewriting this PR to use the latest version please?

ping @Izorkin in case you're interested.

aanderse reviewed (changes requested):
Just a few minor things I'll point out. I only did a very minor 'cosmetic' review of the package. I'm going to leave real review for the big hitters who are familiar with building `mysql`.

orivej commented on pkgs/top-level/all-packages.nix:
```diff
@@ -14594,6 +14594,13 @@ in
+    inherit (darwin) cctools developer_cmds;
+    inherit (darwin.apple_sdk.frameworks) CoreServices;
+    boost = boost169;
+    stdenv = pkgs.clangStdenv;
```
Please use the default boost if mysql compiles with it, and build with the default stdenv.

totten replied:
Re: `boost`: mysql 8.0.16 seemed to be a bit particular about this. The [release notes](https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-16.html) specify the use of boost 1.6.9. With either default `boost` (aka `boost167`) or newer `boost170`, the build fails with similar messages. 

```
-- BOOST_INCLUDE_DIR /nix/store/h6rp8r64zdndyy4hfpd3zbdav4bpqs2x-boost-1.67.0-dev/include
-- LOCAL_BOOST_DIR
-- LOCAL_BOOST_ZIP
-- Could not find (the correct version of) boost.
-- MySQL currently requires boost_1_69_0

CMake Error at cmake/boost.cmake:103 (MESSAGE):
  You can download it with -DDOWNLOAD_BOOST=1 -DWITH_BOOST=<directory>

  This CMake script will look for boost in <directory>.  If it is not there,
  it will download and unpack it (in that directory) for you.

  If you are inside a firewall, you may need to use an https proxy:

  export https_proxy=http://example.com:80

Call Stack (most recent call first):
  cmake/boost.cmake:291 (COULD_NOT_FIND_BOOST)
  CMakeLists.txt:836 (INCLUDE)
```

(I'm pretty sure this error was the same regardless of host platform and of other build options.)

totten replied:
Re: `stdenv` - This was one of the hardest things for me to work out (in no small part b/c I'm inexperienced with this - but I think there's also something wonky going on.)

If one simply changes to use the default `stdenv`, then the build still works on Darwin, but it fails on Linux with:

```
[  7%] Building CXX object extra/icu/source/common/CMakeFiles/icuuc.dir/uvector.cpp.o
[  7%] Building CXX object extra/icu/source/common/CMakeFiles/icuuc.dir/uvectr32.cpp.o
[  7%] Building CXX object extra/icu/source/common/CMakeFiles/icuuc.dir/uvectr64.cpp.o
[  7%] Building CXX object extra/icu/source/common/CMakeFiles/icuuc.dir/wintz.cpp.o
[  7%] Linking CXX static library libicuuc.a
[  7%] Built target icuuc
make: *** [Makefile:152: all] Error 2
builder for '/nix/store/aj50pkqzx2z8z48jbz68f2pwp9xkvnyy-mysql-8.0.16.drv' failed with exit code 2
error: build of '/nix/store/aj50pkqzx2z8z48jbz68f2pwp9xkvnyy-mysql-8.0.16.drv' failed
```

A few data-points/factoids;

- As a general pro-forma thing, I just tried rebasing to use current `nixpkgs@master` (in case stdenv might be diff now), and it still had the same problem.
- When first working on this PR, I wound up trying several flavors of `stdenv` under Darwin+Linux. (`mysql` is a large package, so that took...  patience 😉 ) Setting it to `clangStdenv` was the only thing that worked on both. However, I didn't go in planning/expecting to explore those permutations, so I don't have very systemic notes.
- When first working on the patch, there were a few times where I tried running build commands by hand and used `make -j2` or `make -j3` to speed it up. This manifested new errors - and eventually I realized that those errors didn't happen in a clean build using single-threaded `make`.   Feels like there may be a sequencing issue in the build system.
- The statement `enableParallelBuilding=true` was copied over from `5.7.x.nix`; I didn't register that it might be relevant until @aanderse pointed to its existence.
- I took a shot in the dark and tried permuting `stdenv,enableParallelBuilding` within a Linux system. On `clangStdenv`, the value of `enableParallelBuilding` doesn't seem to have much effect. On the default `stdenv`, it changes the outcome (*although both are failures*). Compare:

    * `Linux, stdenv=clangStdenv, enableParallelBuilding=true`: *Seems to build OK. (Haven't done a new/full run, but the first 10+ min of building seem fine.)*
    * `Linux, stdenv=clangStdenv, enableParallelBuilding=false`: *Seems to build OK. (Haven't done a new/full run, but the first 10+ min of building seem fine.)*
    * `Linux, stdenv=stdenv, enableParallelBuilding=true`: *Fails fairly quickly, with:*

        ```
        [  7%] Building CXX object extra/icu/source/common/CMakeFiles/icuuc.dir/uvector.cpp.o
        [  7%] Building CXX object extra/icu/source/common/CMakeFiles/icuuc.dir/uvectr32.cpp.o
        [  7%] Building CXX object extra/icu/source/common/CMakeFiles/icuuc.dir/uvectr64.cpp.o
        [  7%] Building CXX object extra/icu/source/common/CMakeFiles/icuuc.dir/wintz.cpp.o
        [  7%] Linking CXX static library libicuuc.a
        [  7%] Built target icuuc
        make: *** [Makefile:152: all] Error 2
        builder for '/nix/store/aj50pkqzx2z8z48jbz68f2pwp9xkvnyy-mysql-8.0.16.drv' failed with exit code 2
        error: build of '/nix/store/aj50pkqzx2z8z48jbz68f2pwp9xkvnyy-mysql-8.0.16.drv' failed
        ```

    * `Linux, stdenv=stdenv, enableParallelBuilding=false`: Fails fairly quickly, with:

        ```
        > extern "C" struct rpl_transaction_ctx_service_st {
        >   int (*set_transaction_ctx)(
        >       Transaction_termination_ctx transaction_termination_ctx);
        > } * rpl_transaction_ctx_service;
        > int set_transaction_ctx(
        >     Transaction_termination_ctx transaction_termination_ctx);
        > #include <mysql/service_rpl_transaction_write_set.h>
        > struct Transaction_write_set {
        >   unsigned int m_flags;
        >   unsigned long write_set_size;
        >   unsigned long long *write_set;
        > };
        > extern "C" struct transaction_write_set_service_st {
        >   Transaction_write_set *(*get_transaction_write_set)(
        >       unsigned long m_thread_id);
        > } * transaction_write_set_service;
        > Transaction_write_set *get_transaction_write_set(unsigned long m_thread_id);
        > #include <mysql/service_rules_table.h>
        > #include <string>
        CMake Error at /tmp/nix-build-mysql-8.0.16.drv-0/mysql-8.0.16/cmake/do_abi_check.cmake:86 (MESSAGE):
          ABI check found difference between
          /tmp/nix-build-mysql-8.0.16.drv-0/mysql-8.0.16/include/mysql/services.h.pp
          and /tmp/nix-build-mysql-8.0.16.drv-0/mysql-8.0.16/build/abi_check.out


        make[2]: *** [CMakeFiles/abi_check.dir/build.make:57: CMakeFiles/abi_check] Error 1
        make[1]: *** [CMakeFiles/Makefile2:105: CMakeFiles/abi_check.dir/all] Error 2
        make: *** [Makefile:152: all] Error 2
        builder for '/nix/store/r9gx8an5v5wg5fgja9f7s2mx75zpyhv0-mysql-8.0.16.drv' failed with exit code 2
        error: build of '/nix/store/r9gx8an5v5wg5fgja9f7s2mx75zpyhv0-mysql-8.0.16.drv' failed
        ```

totten replied:
Marking resolved based on @orivej's updates.

totten committed:
mysql80: cleanup pname, url, homepage

totten committed:
mysql80: Remove unnecessary "enableParallelBuilding"

See: https://github.com/NixOS/nixpkgs/pull/65221#discussion_r310383948

totten commented on pkgs/servers/sql/mysql/8.0.x.nix:
```diff
@@ -0,0 +1,94 @@
+
+let
+self = stdenv.mkDerivation rec {
+  name = "mysql-${version}";
```
Thanks @aanderse , that does seem cleaner. :)  I pushed the updates for `pname`, `url`, and `homepage`.

totten commented on pkgs/servers/sql/mysql/8.0.x.nix:
```diff
@@ -0,0 +1,94 @@
+  buildInputs = [ boost libedit libevent lz4 ncurses openssl pkgconfig protobuf readline zlib ]
+     ++ stdenv.lib.optionals stdenv.isDarwin [ perl cctools CoreServices developer_cmds ];
+
+  enableParallelBuilding = true;
```
Thanks for pointing this out. I had just copied it over from `5.7.x.nix`, but it's interesting because (when building 8.0.16 by hand) there were some discrepancies between single-threaded `make` (which I eventually got working) and multi-threaded `make -j2` (which I never got to work completely).

Regardless, since it's not needed, I've removed it.

There may be some connection with the choice of stdenv, so I'll put some more detail in the section about stdenv.

totten committed:
mysql80: 8.0.16 => 8.0.17

Note: "MySQL Router" is a distinct component. Building it fails on macOS 10.12.6 with an error (below); hence, it is skipped.

```
[ 98%] Building CXX object router/src/plugin_info/src/CMakeFiles/mysqlrouter_plugin_info.dir/library_file.cc.o
[ 98%] Building CXX object router/src/plugin_info/src/CMakeFiles/mysqlrouter_plugin_info.dir/main.cc.o
[ 98%] Building CXX object router/src/plugin_info/src/CMakeFiles/mysqlrouter_plugin_info.dir/plugin_info_app.cc.o
[ 98%] Building CXX object router/src/plugin_info/src/CMakeFiles/mysqlrouter_plugin_info.dir/plugin.cc.o
[ 98%] Linking CXX executable ../../../../runtime_output_directory/mysqlrouter_plugin_info
[ 98%] Built target mysqlrouter_plugin_info
Scanning dependencies of target rest_api_lib
[ 98%] Building CXX object router/src/rest_api/src/CMakeFiles/rest_api_lib.dir/rest_api_utils.cc.o
[ 98%] Linking CXX static library librest_api_lib.a
[ 98%] Built target rest_api_lib
Scanning dependencies of target rest_api
[ 98%] Building CXX object router/src/rest_api/src/CMakeFiles/rest_api.dir/rest_api_component.cc.o
In file included from /tmp/nix-build-mysql-8.0.17.drv-0/mysql-8.0.17/router/src/rest_api/src/rest_api_component.cc:30:
/tmp/nix-build-mysql-8.0.17.drv-0/mysql-8.0.17/router/src/rest_api/src/rest_api_plugin.h:90:8: error: 'shared_timed_mutex' is unavailable: introduced in macOS 10.12
  std::shared_timed_mutex rest_api_handler_mutex_;
       ^
/nix/store/3r156gjbpig7xav3crnhffn3msnqmak7-libc++-7.1.0/include/c++/v1/shared_mutex:204:58: note: 'shared_timed_mutex' has been explicitly marked unavailable here
class _LIBCPP_TYPE_VIS _LIBCPP_AVAILABILITY_SHARED_MUTEX shared_timed_mutex
                                                         ^
1 error generated.
make[2]: *** [router/src/rest_api/src/CMakeFiles/rest_api.dir/build.make:63: router/src/rest_api/src/CMakeFiles/rest_api.dir/rest_api_component.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:10908: router/src/rest_api/src/CMakeFiles/rest_api.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
builder for '/nix/store/19c1b8qn0x8fnsa64y4k19r169ifv1b7-mysql-8.0.17.drv' failed with exit code 2
error: build of '/nix/store/19c1b8qn0x8fnsa64y4k19r169ifv1b7-mysql-8.0.17.drv' failed
```

Would skipping it be a regression?  I don't think so: `mysqlrouter` was not
part of `mysql57`, and `mysql80` has not been previously published via
nixpkgs.

totten commented:
> Thanks for deciding to contribute to nixpkgs! It looks like 8.0.17 has been released so when you get a chance would you mind rewriting this PR to use the latest version please?

Alrighty, I've pushed up 585da7881ebe525ef1b6770f3ff093abc4683d05 to use 8.0.17.

There was one snag with 8.0.17 -- the mysql8 source tree includes a bundled copy of `mysqlrouter`, but this isn't building with Darwin and 8.0.17. (See commit message for details.) However, it's basically a new feature and distinct submodule (which wasn't in `mysql57`, `mysql56`, etc), so I suppose there's no major harm in skipping it for the moment?

aanderse commented on pkgs/servers/sql/mysql/8.0.x.nix:
```diff
@@ -45,6 +45,7 @@ self = stdenv.mkDerivation rec {
     "-DWITH_ARCHIVE_STORAGE_ENGINE=yes"
     "-DWITH_BLACKHOLE_STORAGE_ENGINE=yes"
     "-DWITH_FEDERATED_STORAGE_ENGINE=yes"
+    "-DWITH_ROUTER=no" ## mysqlrouter is new addition in 8.0 source-tree, but it currently fails to build on Darwin (tested: macOS 10.12.6 + mysql 8.0.17)
```
Given this is only failing on `darwin` you could make conditional on `stdenv.isDarwin`, if you were so inclined.

totten replied:
If it's dealer's-choice, then I'd lean toward clean "off" since I don't use `mysqlrouter` but do rely on portability between nix's Linux/Darwin environments. But that opinion isn't super-strong - if it's general-practice round here to do a toggle like that, then I'm happy to put the toggle in.

aanderse replied:
dealer's choice it is :+1:

orivej committed:
mysql80: clean up

orivej commented:
I have cleaned it up and consider it ready to merge (as a single squashed commit).

Some notes:
1. `8.0-gc-sections.patch` is not needed because mysql checks if `-fuse-ld=gold` works before enabling it.
2. mysql abi check fails with the default Linux stdenv due to #44530. It is not obvious that we should fix that and it is not fixed yet, so I have included a workaround for mysql.
3. mysqlrouter fails to build on Darwin because it uses `std::shared_timed_mutex` that requires MACOSX_DEPLOYMENT_TARGET to be at least 10.12. I have set it, but disable mysqlrouter anyway because it is difficult to package it properly (it may require changing INSTALL_LAYOUT from STANDALONE to something else), and because it seems possible to package mysqlrouter separately from mysql.

totten commented:
Looks nicer!

I did a local sniff test with the latest revision - built on OSX, launched `mysqld` manually,  connected with `mysql`, and created a dummy DB+table. That all went well. 👍
